R/expr-meta.R

```diff
@@ -67,16 +103,177 @@ expr_meta__as_selector <- function() {
+#' foo <- pl$col("foo")
+#' foo_bar$meta$ne(foo)
+#'
+#' foo_bar2 <- pl$col("foo")$alias("bar")
+#' foo_bar$meta$ne(foo_bar2)
+expr_meta_ne <- function(other) {
+  !self$eq(other)
+}
+
+#' Get a list with the root column name
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' e <- pl$col("foo") * pl$col("bar")
+#' e$meta$root_names()
+#'
+#' e_filter <- pl$col("foo")$filter(pl$col("bar") == 13)
+#' e_filter$meta$root_names()
+#'
+#' e_sum_over <- pl$sum("foo")$over("groups")
+#' e_sum_over$meta$root_names()
+#'
+#' e_sum_slice <- pl$sum("foo")$slice(pl$len() - 10, pl$col("bar"))
+#' e_sum_slice$meta$root_names()
+expr_meta_root_names <- function() {
+  self$`_rexpr`$meta_root_names()
+}
+
+#' Format the expression as a tree
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @param return_as_string Return the tree as a character vector? If `FALSE`
+#' (default), the tree is printed in the console.
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' my_expr <- (pl$col("foo") * pl$col("bar"))$sum()$over(pl$col("ham")) / 2
+#' my_expr$meta$tree_format()
+expr_meta_tree_format <- function(..., return_as_string = FALSE) {
+  wrap({
+    check_dots_empty0(...)
+    out <- self$`_rexpr`$meta_tree_format()
+    if (isTRUE(return_as_string)) {
+      out
+    } else {
+      cat(out)
+    }
+  })
+}
+
+# TODO: add examples with selectors when implemented
+#' Indicate if this expression only selects columns (optionally with aliasing)
+#'
+#' This can include bare columns, column matches by regex or dtype, selectors
+#' and exclude ops, and (optionally) column/expression aliasing.
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @param allow_aliasing If `FALSE` (default), any aliasing is not considered
+#' pure column selection. Set `TRUE` to allow for column selection that also
+#' includes aliasing.
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' e <- pl$col("foo")
+#' e$meta$is_column_selection()
+#'
+#' e <- pl$col("foo")$alias("bar")
+#' e$meta$is_column_selection()
+#'
+#' e$meta$is_column_selection(allow_aliasing = TRUE)
+#'
+#' e <- pl$col("foo") * pl$col("bar")
+#' e$meta$is_column_selection()
+expr_meta_is_column_selection <- function(..., allow_aliasing = FALSE) {
+  wrap({
+    check_dots_empty0(...)
+    self$`_rexpr`$meta_is_column_selection(allow_aliasing)
+  })
+}
+
+#' Indicate if this expression expands to columns that match a regex pattern
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' e <- pl$col("^.*$")$name$prefix("foo_")
+#' e$meta$is_regex_projection()
+expr_meta_is_regex_projection <- function() {
+  wrap({
+    self$`_rexpr`$meta_is_regex_projection()
+  })
+}
+
+#' Pop the latest expression and return the input(s) of the popped expression
+#'
+#' @inherit as_polars_expr return
+#' @examples
+#' e <- pl$col("foo")$alias("bar")
+#' first <- e$meta$pop()
+#' first$meta
+expr_meta_pop <- function() {
+  wrap({
+    out <- self$`_rexpr`$meta_pop()
+    lapply(out, .savvy_wrap_PlRExpr)
```

Comment:
Since this is a list, I think we need to `wrap` the individual elements.
Like these:
https://github.com/eitsupi/neo-r-polars/blob/fd7922069bd0512b78f4c969d62f338a280af958/R/datatypes-classes.R#L115-L121
https://github.com/eitsupi/neo-r-polars/blob/fd7922069bd0512b78f4c969d62f338a280af958/R/dataframe-frame.R#L216-L219

Reply:
That worked, thanks